common/error.hpp

```diff
@@ -13,31 +13,42 @@
 //    limitations under the License.
 
 #include <string>
+#include <system_error>
 #include <type_traits>
 
 #ifndef MENDER_COMMON_ERROR_HPP
 #define MENDER_COMMON_ERROR_HPP
 
 namespace mender::common::error {
 
-template <typename ErrorCodeType>
+template <typename ErrorType>
 class Error {
 public:
-	static_assert(
-		std::is_enum<ErrorCodeType>::value, "Error requires an enum type for error codes");
-	ErrorCodeType error_code;
+	ErrorType error;
 	std::string message;
 
-	Error(const ErrorCodeType &ec, const std::string &msg) {
-		this->error_code = ec;
-		this->message = msg;
-	};
-	Error(const Error &e) {
-		this->error_code = e.error_code;
-		this->message = e.message;
-	};
+	Error() {
```

Comment:
What's the default for an `enum` error code?

Reply:
There's actually two issues here, so let's take one at a time:

* I'm thinking the default is zero, which then becomes the same question as the bool operator above.

* Independently from that, we probably need to explicitly set it to zero, since C++ just uninitialized memory, I believe (I've become too used to Golang here).

Reply:
> * Independently from that, we probably need to explicitly set it to zero, since C++ just uses uninitialized memory, I believe (I've become too used to Golang here).

But we don't have to allow this, right? Can we just mark the default constructor as deleted?

Reply:
Yes, let's use the `NoError` constant instead.